src/sentry/ingest/transaction_clusterer/tree.py

```diff
@@ -0,0 +1,136 @@
+""" Build a directory tree from URL patterns and merge nodes with many siblings.
+
+For example, a blog project might contain transaction names along the lines of:
+
+/users/my-user-name/posts/2022-01-01-that-one-time-i-did-something
+
+We build a tree that looks like this:
+
+/users
+  /my-user-name
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+  /my-user-name2
+    /posts
+      /2022-12-31-i-did-something-too
+  /my-user-name3
+    /posts
+      /2022-07-15-but-what-about-me
+    /settings
+
+As soon as the node /users has more than X children, we merge them:
+
+/users
+  /*
+    /posts
+      /2022-01-01-that-one-time-i-did-something
+      /2022-12-31-i-did-something-too
+      /2022-07-15-but-what-about-me
+    /settings
+
+If /posts now also has more than X children, those are merged as well:
+
+/users
+  /*
+    /posts
+      *
+    /settings
+
+Resulting in the following replacement rules:
+
+/users/*/posts/*  # To replace both identifiers
+/users/*          # For URLs with only one identifier, e.g. /users/*/settings
+
+"""
+
+from collections import UserDict, defaultdict
+from typing import Iterable, List, Optional, Union
+
+from typing_extensions import TypeAlias
+
+from .base import Clusterer, ReplacementRule
+
+__all__ = ["TreeClusterer"]
+
+
+class Merged:
+    pass
+
+
+#: Symbol representing a merged node
+MERGED = Merged()
+
+#: Separator by which we build the tree
+SEP = "/"
+
+
+# TODO: Metrics for tree size, spans for merge, etc.
+
+
+class TreeClusterer(Clusterer):
+    def __init__(self, *, merge_threshold: int) -> None:
+        self._merge_threshold = merge_threshold
+        self._graph = Node()
+
+    def add_input(self, transaction_names: Iterable[str]) -> None:
+        for tx_name in transaction_names:
+            parts = tx_name.split(SEP)
+            node = self._graph
+            for part in parts:
+                node = node.setdefault(part, Node())
+
+    def get_rules(self) -> List[ReplacementRule]:
+        """Merge high-cardinality nodes in the graph and extract rules"""
+        self._graph.merge(self._merge_threshold)
+        # Generate exactly 1 rule for every merge
+        rule_paths = [path for path in self._graph.paths() if path[-1] is MERGED]
+
+        # TODO: Warn if top level collapse
+
+        # Sort by path length, descending (most specific rule first)
+        rule_paths.sort(key=len, reverse=True)
+
+        return [self._build_rule(path) for path in rule_paths]
+
+    @staticmethod
+    def _build_rule(path: List["Edge"]) -> ReplacementRule:
+        return ReplacementRule(SEP.join(["*" if isinstance(key, Merged) else key for key in path]))
```

Comment:
Here we can make a distinction between a merged node, and merged and terminal node. In the former, we add `*` to the rule; in the latter, we add `**` instead. This improves the glob for the transactions where they don't need a `**` somewhere in the middle. I haven't thought about how to approach these cases, so there may be a possibility too.

Reply:
See comment below for further simplification.